### PR TITLE
Async job schedule

### PR DESCRIPTION
**Pull request by zhoney**

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/ServerInfoManager.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+    private Iterator<HugeServerInfo> serverInfos(Map<String, Object> conditions,
+                                                 long limit, String page) {
+        ConditionQuery query = new ConditionQuery(HugeType.VERTEX);
+        if (page != null) {
+            query.page(page);
+        }
+        VertexLabel vl = this.graph.graph().vertexLabel(HugeServerInfo.P.SERVER);
+
+
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/ServerInfoManager.java:203]: two empty lines

File: hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/HugeServerInfo.java [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+                                     .properties(properties)
+                                     .useCustomizeStringId()
+                                     .nullableKeys(P.ROLE, P.LOAD,
+                                                   P.UPDATE_TIME)
+                                     .enableLabelIndex(true)
+                                     .build();
+            this.graph.schemaTransaction().addVertexLabel(label);
+
+            // Create index
+            this.createIndexLabel(label, P.ROLE);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/HugeServerInfo.java:248]: seems no need to query by role

File: hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/HugeServerInfo.java [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+import com.baidu.hugegraph.type.define.Cardinality;
+import com.baidu.hugegraph.type.define.DataType;
+import com.baidu.hugegraph.type.define.GraphRole;
+import com.baidu.hugegraph.type.define.SerialEnum;
+import com.baidu.hugegraph.util.DateUtil;
+import com.baidu.hugegraph.util.E;
+
+public class HugeServerInfo {
+
+    public static final int MAX_LOAD = 1000;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/HugeServerInfo.java:53]: report MAX_LOAD by itself and add maxLoad field

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/TaskManager.java [hunk 4]
```diff
@@ -59,13 +68,21 @@ private TaskManager(int pool) {
         this.taskExecutor = ExecutorUtil.newFixedThreadPool(pool, TASK_WORKER);
         // For save/query task state, just one thread is ok
         this.dbExecutor = ExecutorUtil.newFixedThreadPool(1, TASK_DB_WORKER);
+        // For schedule task to run, just one thread is ok
+        this.taskScheduler = ExecutorUtil.newScheduledThreadPool(
+                             1, TASK_SCHEDULER);
+        // Start after 10s waiting for HugeGraphServer startup
+        this.taskScheduler.scheduleWithFixedDelay(this::periodicJob,
+                                                  10L, 3, TimeUnit.SECONDS);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/TaskManager.java:76]: define 3 as const var SCHEDULE_PERIOD

File: hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/HugeServerInfo.java [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+    public static final long EXPIRED_INTERVAL = 5000L;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/HugeServerInfo.java:54]: SCHEDULE_PERIOD * 3

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/job/TaskAPI.java [hunk 6]
```diff
@@ -161,11 +161,11 @@ public void delete(@Context GraphManager manager,
 
         TaskScheduler scheduler = graph(manager, graph).taskScheduler();
         HugeTask<?> task = scheduler.task(IdGenerator.of(id));
-        if (!task.completed() && scheduler.cancel(task)) {
+        if (!task.completed() &&
+            (scheduler.cancel(task) || task.cancelling())) {
```
[the remark this answers is not in the document]
Comment (zhoney) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/api/job/TaskAPI.java:165]: `scheduler.cancel(task)` will change task status to `CANCELLING`

File: hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/HugeServerInfo.java (hunk 2 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/HugeServerInfo.java:248]: query by role in `initServerInfo` of `com/baidu/hugegraph/cluster/ServerInfoManager.java`

File: hugegraph-api/src/main/java/com/baidu/hugegraph/config/ServerOptions.java [hunk 7]
```diff
@@ -57,6 +57,15 @@ public static synchronized ServerOptions instance() {
                     "server-1"
             );
 
+    public static final ConfigOption<String> SERVER_ROLE =
+            new ConfigOption<>(
+                    "server.role",
+                    "The role of nodes in the cluster, available type are " +
+                    "[master, worker]",
+                    disallowEmpty(),
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/config/ServerOptions.java:65]: use allowValues

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java [hunk 8]
```diff
@@ -47,16 +47,15 @@
     private EventListener storeEventListener;
     private EventListener cacheEventListener;
 
-    private final Map<HugeType, Boolean> cachedTypes;
+    private static final Map<HugeGraphParams, CachedTypes> CACHED_TYPES =
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java:50]: Map<String, CachedTypes>

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/TaskManager.java [hunk 9]
```diff
@@ -163,6 +207,37 @@ public int pendingTasks() {
         return size;
     }
 
+    protected void notifyNewTask(HugeTask<?> task) {
+        this.schedulerExecutor.submit(this::scheduleOrExecuteJob);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/TaskManager.java:211]: add some comments

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/TaskManager.java [hunk 10]
```diff
@@ -163,6 +207,37 @@ public int pendingTasks() {
...
+            for (Map.Entry<HugeGraphParams, TaskScheduler> entry :
+                 this.schedulers.entrySet()) {
+                ServerInfoManager server = entry.getKey().serverManager();
+                StandardTaskScheduler scheduler = (StandardTaskScheduler)
+                                                  entry.getValue();
+
+                // Update server heartbeat
+                server.heartbeat();
+
+                // Master schedule tasks not found suitable server when created
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/TaskManager.java:225]: improve "not found suitable server when created"

File: hugegraph-core/src/main/java/com/baidu/hugegraph/StandardHugeGraph.java [hunk 11]
```diff
@@ -267,6 +267,8 @@ public void truncateBackend() {
 
         this.storeProvider.truncate();
         this.storeProvider.initSystemInfo(this);
+        this.serverStarted(this.serverManager.serverId().asString(),
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/StandardHugeGraph.java:270]: why not change the parameter type

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java [hunk 12]
```diff
@@ -47,16 +47,15 @@
     private EventListener storeEventListener;
     private EventListener cacheEventListener;
 
-    private final Map<HugeType, Boolean> cachedTypes;
+    private static final Map<String, CachedTypes> CACHED_TYPES =
+            new ConcurrentHashMap<>();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java:51]: prefer align with CACHED_TYPES

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java [hunk 13]
```diff
@@ -252,9 +259,12 @@ protected void removeSchema(SchemaElement schema) {
...
                 }
-                this.cachedTypes.putIfAbsent(type, true);
+                this.cachedTypes().putIfAbsent(type, true);
             }
             return results;
         }
     }
+
+    private static class CachedTypes
+            extends ConcurrentHashMap<HugeType, Boolean> {}
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/cache/CachedSchemaTransaction.java:269]: align with class

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeTask.java [hunk 14]
```diff
@@ -642,28 +636,14 @@ private void checkPropertySize(int propertyLength, String propertyName) {
...
-                throw (RuntimeException) cause;
-            }
-            throw new HugeException("Async task failed with error: %s",
-                                    cause, cause.getMessage());
         } catch (Exception e) {
-            throw new HugeException("Async task failed with error: %s",
-                                    e, e.getMessage());
-        } finally {
-            ((StandardTaskScheduler) task.scheduler()).remove(id);
+            // ignore
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeTask.java:646]: don't ignore exception here

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeTask.java [hunk 15]
```diff
@@ -642,28 +636,14 @@ private void checkPropertySize(int propertyLength, String propertyName) {
...
-        HugeTask task = null;
+        HugeTask task;
         try {
-            Id id = this.id();
             do {
-                task = this.scheduler().task(id);
-                if (task.status.code() > TaskStatus.QUEUED.code()) {
-                    task.get();
-                }
+                task = this.scheduler().task(this.id());
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeTask.java:642]: should sleep between tow times of task-read

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java [hunk 16]
```diff
@@ -322,12 +325,13 @@ private void unlistenChanges() {
         HugeServerInfo server;
         do {
             Iterator<HugeServerInfo> servers = this.serverManager()
-                                                   .serverInfos(100L, page);
+                                                   .serverInfos(10L, page);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java:328]: add const var

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java [hunk 17]
```diff
@@ -388,9 +392,10 @@ protected ServerInfoManager serverManager() {
         return this.graph.serverManager();
     }
 
-    protected void remove(Id id) {
+    public void remove(Id id) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java:395]: keep protected

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexLabelCoreTest.java [hunk 18]
```diff
@@ -900,13 +904,22 @@ public void testRemoveVertexLabelUsedByEdgeLabel() {
...
+        Assert.assertContains("Not allowed to remove vertex label " +
+                              "'person' because the edge label 'write' " +
+                              "still link with it", task.result());
+
+        id = schema.vertexLabel("book").remove();
+        sleepAWhile(1000L);
+        task = scheduler.task(id);
+        Assert.assertEquals(TaskStatus.FAILED, task.status());
+        Assert.assertContains("Not allowed to remove vertex label 'book' " +
+                              "because the edge label 'write' still link with" +
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/VertexLabelCoreTest.java:921]: wrap line after "link "

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/TaskCoreTest.java [hunk 19]
```diff
@@ -565,7 +575,38 @@ public void testGremlinJobAndRestore() throws Exception {
...
+        @Override
+        public V call() throws Exception {
+            Thread.sleep(SLEEP_TIME);
+            return null;
+        }
+
+        @Override
+        public void done() {
+            StandardTaskScheduler scheduler = Whitebox.getInternalState(
+                                              this.task(), "scheduler");
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/TaskCoreTest.java:608]: why not call this.task().scheduler()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java [hunk 20]
```diff
@@ -246,16 +246,19 @@ private void unlistenChanges() {
...
     @Override
-    public <V> void cancel(HugeTask<V> task) {
+    public synchronized <V> void cancel(HugeTask<V> task) {
         E.checkArgumentNotNull(task, "Task can't be null");
         if (!this.serverManager().master()) {
             return;
         }
         if (!task.completed()) {
             // The task scheduled to workers, waiting for worker cancel
+            task = this.task(task.id());
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java:256]: why read again

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java [hunk 21]
```diff
@@ -193,12 +204,26 @@ private void unlistenChanges() {
...
+        this.save(task);
+
+        // Notify master server to schedule and execute immediately
+        TaskManager.instance().notifyNewTask(task);
+
+        return task;
     }
 
-    private <V> Future<?> submitTask(HugeTask<V> task) {
+    public  <V> Future<?> submitTask(HugeTask<V> task) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java:226]: keep private

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/TaskManager.java [hunk 22]
```diff
@@ -26,27 +26,37 @@
...
 
 public final class TaskManager {
 
+    private static final Logger LOG = Log.logger(TaskManager.class);
+
     public static final String TASK_WORKER = "task-worker-%d";
     public static final String TASK_DB_WORKER = "task-db-worker-%d";
+    public static final String TASK_SCHEDULER = "task-scheduler-%d";
 
+    public static final int SCHEDULE_PERIOD = 3;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/TaskManager.java:51]: add comment: Unit second

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/MultiGraphsTest.java [hunk 23]
```diff
@@ -325,8 +329,17 @@ public void testCreateGraphsWithMultiDisksForRocksDB() {
...
         }
         ((BaseConfiguration) config).setDelimiterParsingDisabled(true);
+        String backend = config.getString(CoreOptions.BACKEND.name());
+        String data = config.getString(RocksDBOptions.DATA_PATH.name());
+        String wal = config.getString(RocksDBOptions.DATA_PATH.name());
         for (String graphName : graphNames) {
             config.setProperty(CoreOptions.STORE.name(), graphName);
+            if (backend.equals("rocksdb")) {
+                String dataPath = data + "/" + graphName;
+                config.setProperty(RocksDBOptions.DATA_PATH.name(), dataPath);
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/MultiGraphsTest.java:339]: prefer set DATA_PATH before openGraphs() in each test

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java [hunk 24]
```diff
@@ -550,13 +700,14 @@ public void initSchema() {
...
                                                    P.UPDATE, P.INPUT, P.RESULT,
-                                                   P.DEPENDENCIES)
+                                                   P.DEPENDENCIES, P.SERVER)
                                      .enableLabelIndex(true)
                                      .build();
             this.params().schemaTransaction().addVertexLabel(label);
 
             // Create index
             this.createIndexLabel(label, P.STATUS);
+            this.createIndexLabel(label, P.SERVER);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java:710]: unneeded index

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java [hunk 25]
```diff
@@ -226,31 +246,161 @@ private void unlistenChanges() {
...
+                    this.submitTask(task);
+                }
+            }
+            page = PageInfo.pageInfo(tasks);
+        } while (page != null);
+    }
+
+    private synchronized <V> HugeServerInfo pickWorker(HugeTask<V> task) {
+        HugeServerInfo master = null;
+        HugeServerInfo minServer = null;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java:318]: minServer -- the meaning is incomprehensible

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java [hunk 26]
```diff
@@ -193,12 +204,26 @@ private void unlistenChanges() {
...
     @Override
     public <V> Future<?> schedule(HugeTask<V> task) {
+        if (!this.serverManager().master()) {
+            throw new HugeException("The worker can't schedule task");
+        }
         E.checkArgumentNotNull(task, "Task can't be null");
+
         task.status(TaskStatus.QUEUED);
-        return this.submitTask(task);
+        if (task.callable() instanceof EphemeralJob) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java:213]: add some comments

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java [hunk 27]
```diff
@@ -162,6 +170,7 @@ private void unlistenChanges() {
     public <V> void restoreTasks() {
         boolean supportsPaging = this.graph().backendStoreFeatures()
                                              .supportsQueryByPage();
+        Id server = this.serverManager().serverId();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java:173]: rename to selfServer

File: hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java [hunk 28]
```diff
@@ -114,7 +114,7 @@ public static synchronized CoreOptions instance() {
                     "Timeout in seconds for waiting for the task to complete," +
                     "such as when truncating or clearing the backend.",
                     rangeInt(0L, Long.MAX_VALUE),
-                    10L
+                    30L
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java:117]: too much time that client may time out, 20s is ok

File: hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/ServerInfoManager.java [hunk 29]
```diff
@@ -0,0 +1,247 @@
...
+                                    "cluster", existed.id());
+                }
+                page = PageInfo.pageInfo(servers);
+            } while (page != null);
+        }
+
+        HugeServerInfo serverInfo = new HugeServerInfo(server, nodeRole);
+        serverInfo.maxLoad(this.calcMaxLoad());
+        this.serverId = serverInfo.id();
+        this.serverRole = nodeRole;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/ServerInfoManager.java:128]: move serverId and serverRole assignment to the begin of this method after refactor initServerInfo params

File: hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/ServerInfoManager.java [hunk 30]
```diff
@@ -0,0 +1,247 @@
...
+        E.checkArgument(existed == null || !existed.alive(),
+                        "The server with name '%s' already in cluster",
+                        server);
+        E.checkArgument(role != null && !role.isEmpty(),
+                        "The server role can't be null or empty");
+        NodeRole nodeRole = NodeRole.valueOf(role.toUpperCase());
+        if (nodeRole.master()) {
+            String page = PAGE_NONE;
+            do {
+                Iterator<HugeServerInfo> servers = this.serverInfos(10L, page);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/ServerInfoManager.java:113]: add const var

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/TaskCoreTest.java (hunk 19 above)
Comment (zhoney) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/core/TaskCoreTest.java:608]: HugeTask doesn't have scheduler()
Already change to this.graph().taskScheduler()

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/MultiGraphsTest.java (hunk 23 above)
Comment (javeme) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/core/MultiGraphsTest.java:340]: address it

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeTask.java [hunk 31]
```diff
@@ -592,18 +637,10 @@ private void checkPropertySize(int propertyLength, String propertyName) {
...
-            this.get();
-            assert this.completed();
-        } catch (ExecutionException e) {
-            Throwable cause = e.getCause();
-            if (cause instanceof RuntimeException) {
-                throw (RuntimeException) cause;
-            }
-            throw new HugeException("Async task failed with error: %s",
-                                    cause, cause.getMessage());
+            this.scheduler().waitUntilTaskCompleted(this.id(), 20L, 10);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeTask.java:640]: 10L

Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeTask.java:640]: seems can delete syncWait() method

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java [hunk 32]
```diff
@@ -226,31 +250,160 @@ private void unlistenChanges() {
...
+            HugeTask<V> task;
+            while (tasks.hasNext()) {
+                task = tasks.next();
+                if (task.server() != null) {
+                    // Skip if already scheduled
+                    continue;
+                }
+                HugeServerInfo server = this.pickWorker(task);
+                if (server == null) {
+                    LOG.debug("The master can not find suitable server to " +
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java:283]: "can't"

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/TaskCoreTest.java [hunk 33]
```diff
@@ -325,6 +316,7 @@ public void testGremlinJobWithFailure() throws TimeoutException {
                .job(new GremlinJob());
         HugeTask<Object> task = builder.schedule();
         scheduler.waitUntilTaskCompleted(task.id(), 10);
+        task = scheduler.task(task.id());
```
Comment (javeme) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/TaskCoreTest.java:319]: task = scheduler.waitUntilTaskCompleted()

**Comment by javeme** [comment]
test errors:
```
[ERROR] Tests run: 288, Failures: 0, Errors: 2, Skipped: 1, Time elapsed: 289.868 s <<< FAILURE! - in com.baidu.hugegraph.unit.UnitTestSuite
[ERROR] testResetCachedAllIfReachedCapacity(com.baidu.hugegraph.unit.cache.CachedSchemaTransactionTest)  Time elapsed: 0.137 s  <<< ERROR!
java.lang.RuntimeException: Unable to get internal state on field 'cachedTypes' of com.baidu.hugegraph.backend.cache.CachedSchemaTransaction@2d7ee91c
	at com.baidu.hugegraph.unit.cache.CachedSchemaTransactionTest.testResetCachedAllIfReachedCapacity(CachedSchemaTransactionTest.java:186)
Caused by: java.lang.RuntimeException: Not declared field 'cachedTypes' in class 'Object'
	at com.baidu.hugegraph.unit.cache.CachedSchemaTransactionTest.testResetCachedAllIfReachedCapacity(CachedSchemaTransactionTest.java:186)
[ERROR] com.baidu.hugegraph.unit.core.SecurityManagerTest  Time elapsed: 0.402 s  <<< ERROR!
com.baidu.hugegraph.HugeException: The worker can't schedule task
	at com.baidu.hugegraph.unit.core.SecurityManagerTest.runGremlinJob(SecurityManagerTest.java:311)
	at com.baidu.hugegraph.unit.core.SecurityManagerTest.init(SecurityManagerTest.java:62)
```

File: hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/HugeServerInfo.java [hunk 34]
```diff
@@ -176,9 +183,8 @@ public static HugeServer fromVertex(Vertex vertex) {
         return server;
     }
 
-    public <V> boolean suitableFor(HugeTask<V> task, long now) {
-        if (this.updateTime.getTime() + 5000L < now ||
-            this.load() + task.load() > MAX_LOAD) {
+    public <V> boolean suitableFor(HugeTask<V> task) {
+        if (this.alive() || this.load() + task.load() > MAX_LOAD) {
```
Comment (houzhizhen) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/HugeServerInfo.java]: seem contrary to common sense, it should accept new task if it is alive

File: hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/ServerInfoManager.java [hunk 35]
```diff
@@ -93,13 +94,29 @@ public boolean close() {
     public void initServerInfo(String server, String role) {
         E.checkArgument(server != null && !server.isEmpty(),
                         "The server name can't be null or empty");
+        HugeServerInfo existed = this.serverInfo(server);
+        E.checkArgument(existed == null || !existed.alive(),
+                        "The server with name '%s' already in cluster",
```
Comment (houzhizhen) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/ServerInfoManager.java:107]: The exception information may not correct.

File: hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/ServerInfoManager.java [hunk 36]
```diff
@@ -93,13 +94,29 @@ public boolean close() {
...
-        this.serverId = hugeServer.id();
+        if (graphRole.master()) {
+            Iterator<HugeServerInfo> servers = this.serverInfos(ImmutableMap.of(
+                                               HugeServerInfo.P.ROLE,
+                                               GraphRole.MASTER.code()),
+                                               1, PAGE_NONE);
+            if (servers.hasNext()) {
+                existed = servers.next();
+                E.checkArgument(!existed.alive(),
+                                "Already existed master '%s' in current " +
```
Comment (houzhizhen) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/cluster/ServerInfoManager.java]: Exception information may not correct.

**Comment by zhoney** [comment]
```
[ERROR] testList(com.baidu.hugegraph.api.EdgeLabelApiTest)  Time elapsed: 0.286 s  <<< FAILURE!
java.lang.AssertionError: Response with status 400 and content {"exception":"class java.lang.IllegalArgumentException","message":"Undefined source vertex label 'person' in edge label 'created'","cause":""} expected:<201> but was:<400>
	at com.baidu.hugegraph.api.EdgeLabelApiTest.testList(EdgeLabelApiTest.java:115)
[ERROR] testBatchUpdate(com.baidu.hugegraph.api.EdgeApiTest)  Time elapsed: 0.21 s  <<< ERROR!
com.baidu.hugegraph.HugeException: Failed to get vertex id: {"exception":"class java.lang.IllegalArgumentException","message":"Undefined vertex label: 'person'","cause":""}
	at com.baidu.hugegraph.api.EdgeApiTest.testBatchUpdate(EdgeApiTest.java:65)
```

**Comment by zhoney** [comment]
hbase
```
[ERROR] testGremlinJobAndCancel(com.baidu.hugegraph.core.TaskCoreTest)  Time elapsed: 3.219 s  <<< FAILURE!
java.lang.AssertionError: expected:<CANCELLED> but was:<CANCELLING>
	at com.baidu.hugegraph.core.TaskCoreTest.testGremlinJobAndCancel(TaskCoreTest.java:457)
[ERROR] testGremlinJobAndRestore(com.baidu.hugegraph.core.TaskCoreTest)  Time elapsed: 0.066 s  <<< ERROR!
com.baidu.hugegraph.HugeException: Failed to update/query TaskStore: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Can't delete incomplete task '1760' in status CANCELLING
	at com.baidu.hugegraph.core.TaskCoreTest.setup(TaskCoreTest.java:64)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Can't delete incomplete task '1760' in status CANCELLING
	at com.baidu.hugegraph.core.TaskCoreTest.setup(TaskCoreTest.java:64)
Caused by: java.lang.IllegalStateException: Can't delete incomplete task '1760' in status CANCELLING
```

**Comment by javeme** [comment]
```
[ERROR] Tests run: 289, Failures: 0, Errors: 2, Skipped: 1, Time elapsed: 277.599 s <<< FAILURE! - in com.baidu.hugegraph.unit.UnitTestSuite
[ERROR] com.baidu.hugegraph.unit.core.SecurityManagerTest  Time elapsed: 0.403 s  <<< ERROR!
com.baidu.hugegraph.HugeException: Failed to update/query server info: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Invalid PageState 'null'
	at com.baidu.hugegraph.unit.core.SecurityManagerTest.loadGraph(SecurityManagerTest.java:330)
	at com.baidu.hugegraph.unit.core.SecurityManagerTest.init(SecurityManagerTest.java:63)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: Invalid PageState 'null'
	at com.baidu.hugegraph.unit.core.SecurityManagerTest.loadGraph(SecurityManagerTest.java:330)
	at com.baidu.hugegraph.unit.core.SecurityManagerTest.init(SecurityManagerTest.java:63)
Caused by: java.lang.IllegalStateException: Invalid PageState 'null'
[ERROR] com.baidu.hugegraph.unit.core.SecurityManagerTest  Time elapsed: 0.404 s  <<< ERROR!
java.lang.NullPointerException
	at com.baidu.hugegraph.unit.core.SecurityManagerTest.clear(SecurityManagerTest.java:71)

```

**Comment by zhoney** [comment]
```
[ERROR] testCreateMultiGraphs(com.baidu.hugegraph.core.MultiGraphsTest)  Time elapsed: 96.731 s  <<< FAILURE!
java.lang.AssertionError: java.lang.IllegalStateException: Ensure tx closed in all threads when closing graph 'g12345678901234567890123456789012345678901234567'
	at com.baidu.hugegraph.core.MultiGraphsTest.destoryGraphs(MultiGraphsTest.java:344)
	at com.baidu.hugegraph.core.MultiGraphsTest.testCreateMultiGraphs(MultiGraphsTest.java:62)
[ERROR] testCopySchemaWithMultiGraphsWithConflict(com.baidu.hugegraph.core.MultiGraphsTest)  Time elapsed: 96.539 s  <<< FAILURE!
java.lang.AssertionError: java.lang.IllegalStateException: Ensure tx closed in all threads when closing graph 'schema_g2'
	at com.baidu.hugegraph.core.MultiGraphsTest.destoryGraphs(MultiGraphsTest.java:344)
	at com.baidu.hugegraph.core.MultiGraphsTest.testCopySchemaWithMultiGraphsWithConflict(MultiGraphsTest.java:209)
```

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java [hunk 37]
```diff
@@ -293,12 +295,16 @@ private void unlistenChanges() {
...
+            if (page != null) {
+                page = PageInfo.pageInfo(tasks);
+            }
         } while (page != null);
     }
 
     protected <V> void executeTasksForWorker(Id server) {
-        String page = PageInfo.PAGE_NONE;
+        boolean supportsPaging = this.graph().backendStoreFeatures()
+                                     .supportsQueryByPage();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java:306]: define a field or method supportsPaging

File: hugegraph-api/src/main/java/com/baidu/hugegraph/config/ServerOptions.java [hunk 38]
```diff
@@ -57,6 +57,15 @@ public static synchronized ServerOptions instance() {
                     "server-1"
             );
 
+    public static final ConfigOption<String> SERVER_ROLE =
+            new ConfigOption<>(
+                    "server.role",
+                    "The role of nodes in the cluster, available types are " +
+                    "[master, worker]",
+                    allowValues("master", "worker"),
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/config/ServerOptions.java:65]: add "computer"

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeTask.java [hunk 39]
```diff
@@ -591,19 +636,22 @@ private void checkPropertySize(int propertyLength, String propertyName) {
     }
 
     public void syncWait() {
+        long timeout = ((HugeConfig) this.scheduler.graph().configuration())
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeTask.java:639]: not supported by auth graph

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeTask.java [hunk 40]
```diff
@@ -591,19 +636,22 @@ private void checkPropertySize(int propertyLength, String propertyName) {
...
-            this.get();
-            assert this.completed();
-        } catch (ExecutionException e) {
-            Throwable cause = e.getCause();
-            if (cause instanceof RuntimeException) {
-                throw (RuntimeException) cause;
-            }
-            throw new HugeException("Async task failed with error: %s",
-                                    cause, cause.getMessage());
+            this.scheduler().waitUntilTaskCompleted(this.id(), timeout, 10L);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/HugeTask.java:642]: can read timeout in scheduler.waitUntilTaskCompleted()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/TaskManager.java [hunk 41]
```diff
@@ -114,18 +151,49 @@ private void closeTaskTx(HugeGraphParams graph) {
...
+        final Callable<Void> closeTx = () -> {
+            // Do close-tx for current thread
+            graph.closeTx();
+            // Let other threads run
+            Thread.yield();
+            return null;
+        };
+
+        try {
+            this.schedulerExecutor.submit(closeTx);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/TaskManager.java:164]: ensure completed, call this.schedulerExecutor.submit(closeTx).get()

File: hugegraph-api/src/main/java/com/baidu/hugegraph/config/ServerOptions.java [hunk 42]
```diff
@@ -62,7 +62,7 @@ public static synchronized ServerOptions instance() {
                     "server.role",
                     "The role of nodes in the cluster, available types are " +
                     "[master, worker]",
-                    allowValues("master", "worker"),
+                    allowValues("master", "worker", "computer"),
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/config/ServerOptions.java:65]: also update line 64

**Comment by zhoney** [comment]
异步任务等待10ms时，CI测试时间每个延长3分钟左右

image

**Comment by javeme** [comment]
improved by https://github.com/hugegraph/hugegraph/pull/1079:
image

**Pull request merged**